core/src/main/java/org/apache/iceberg/BaseRewriteManifests.java

```diff
@@ -312,33 +311,58 @@ long getManifestTargetSizeBytes() {
   }
 
   class WriterWrapper {
-    private ManifestWriter writer;
-
-    synchronized void addEntry(ManifestEntry entry) {
-      if (writer == null) {
-        writer = newWriter();
-      } else if (writer.length() >= getManifestTargetSizeBytes()) {
-        close();
-        writer = newWriter();
+    private final Map<Integer, ManifestWriter> manifestWritersBySpecId = Collections.synchronizedMap(new HashMap<>());
+
+    synchronized void addEntry(ManifestEntry entry, int partitionSpecId) {
+      ManifestWriter manifestWriter = manifestWritersBySpecId.get(partitionSpecId);
+      if (manifestWriter == null) {
+        manifestWriter = addAndReturnNewWriter(partitionSpecId);
+      } else if (manifestWriter.length() >= getManifestTargetSizeBytes()) {
+        close(partitionSpecId);
+        manifestWriter = addAndReturnNewWriter(partitionSpecId);
       }
-      writer.existing(entry);
+      manifestWriter.existing(entry);
     }
 
-    private ManifestWriter newWriter() {
-      return new ManifestWriter(spec, manifestPath(manifestSuffix.getAndIncrement()), snapshotId());
+    private ManifestWriter addAndReturnNewWriter(int partitionSpecId) {
```

Comment:
I would consider splitting this into two methods and calling them as necessary:

```
private ManifestWriter newWriter(int partitionSpecId) {
  PartitionSpec spec = specsById.get(partitionSpecId);
  return new ManifestWriter(spec, manifestPath(manifestSuffix.getAndIncrement()), snapshotId());
}
```

```
private void addManifestWriter(int partitionSpecId, ManifestWriter manifestWriter) {
  if (manifestWritersBySpecId.contains(partitionSpecId)) {
    // throw some exception as this should not happen (will leak resources otherwise)
  }
  manifestWritersBySpecId.put(partitionSpecId, manifestWriter);
}
```

Reply:
will do, should be good.

Reply:
@aokolnychyi 
I also though so, but if we make two methods, there will be duplicate calls to both the methods at this place, and also these methods are not being used from other places.

`      if (manifestWriter == null) {
        manifestWriter = addAndReturnNewWriter(partitionSpecId);
      } else if (manifestWriter.length() >= getManifestTargetSizeBytes()) {
        close(partitionSpecId);
        manifestWriter = addAndReturnNewWriter(partitionSpecId); 
}`

please see, do we still need 2 different methods.

Reply:
What about the following then?

```
private ManifestWriter newWriter(int specId) {
  if (manifestWritersBySpecId.contains(specId)) {
    // throw some exception as this should not happen (will leak resources otherwise)
  }

  PartitionSpec spec = specsById.get(specId);
  OutputFile newFile = manifestPath(manifestSuffix.getAndIncrement());
  ManifestWriter manifestWriter = new ManifestWriter(spec, newFile, snapshotId());

  manifestWritersBySpecId.put(specId, manifestWriter);

  return manifestWriter;
}
```

Reply:
thanks @aokolnychyi.

refactored method name.

wondering do we need to add this check :)  (Hopefully not missing something)
`if (manifestWritersBySpecId.contains(specId)) {
    // throw some exception as this should not happen (will leak resources otherwise)
  }`

as `newWriter` ( adds to `manifestWritersBySpecId` ) is called from other **synchronized** method, and one which remove writer from  is `manifestWritersBySpecId` `close()` and that is also **synchronized**.
So, state of `manifestWritersBySpecId` is protected.

Reply:
I know that it is OK to use `newWriter` right now but we rely on the correct logic in other methods. That's, probably, fine but it felt safer to have this check to be sure we don't replace a writer that hasn't been closed yet (if something changes later on in the code).

Also, I think it would be slightly more readable to have args to `ManifestWriter` as separate variables like in the snippet above.

Reply:
>I know that it is OK to use newWriter right now but we rely on the correct logic in other methods. That's, probably, fine but it felt safer to have this check to be sure we don't replace a writer that hasn't been closed yet (if something changes later on in the code).
Also, I think it would be slightly more readable to have args to ManifestWriter as separate variables like in the snippet above.

make sense. let me take care.

Reply:
this is also done.